Improve Nomad Docs (#660)
openHPI/poseidon: docs/nomad_usage.md

When a user then executes their code, Poseidon copies the code into the container and executes it.

### Nomad Restarts

When the Nomad Servers or Agents restart, running executions can be terminated.  
For agents, it depends on whether the runner allocation is placed on the restarted agent.
For servers, it depends both on the role - if the restarted server is the cluster leader - and on whether Poseidon is connected to the restarted server, e.g. due to DNS Resolving.

> Poseidon can be connected to the server either for individual execution or for receiving the Nomad Event stream.

Reviewer: ```suggestion
Poseidon can be connected to the server either for individual execution or for receiving the Nomad Event Stream.
```

The following table lists the behavior for restarts of Nomad Servers depending on its role and on whether Poseidon is connected to it (via DNS Resolution).

| Role     | DNS Resolves | | WebSocket Problem? | Event Stream Problem? | 
|----------|--------------|-|--------------------|-----------------------|
| Leader   | Yes          | | problematic        | problematic           |
| Leader   | No           | | problematic        | fine                  |
| Follower | Yes          | | problematic        | problematic           |
| Follower | No           | | fine               | fine                  |

Such restarts lead to problems with either individual WebSocket connections of executions or the Nomad Event Stream.
When the Nomad Event Stream is aborted, Poseidon tries to reestablish it. Once it succeeds in doing so, all environments and runners are recovered from Nomad.

> In the case of Nomad Agent restarts the WebSocket connection of a running execution aborts.

Reviewer: unsure 

```suggestion
In the case of Nomad Agent restarts, the WebSocket connection of a running execution aborts.
```

> Furthermore, when also Docker of the Nomad Agent is restarted, the containers are recreated.

Reviewer: This is part of our custom setting through the `PartOf` directive in Nomad, isn't it? Is said directive documented here?

Author: Barely, since a restart of Nomad does not trigger a restart of Docker (only vice versa).

Reviewer: Mh, I thought so. See [this setting](https://lab.xikolo.de/codeocean/codeocean-terraform/blob/31f962df429c27f7b69c36984589284e59b03bb2/ansible/playbooks/roles/nomad/tasks/main.yml#L76-L87) for details.

Author: <details><summary>Logs from restart test</summary>
<p>

```
cat /etc/systemd/system/nomad.service.d/override.conf
[Service]
TimeoutStopSec = 20
ExecStartPost = systemctl restart nomad.eligible.service
[Unit]
PartOf = docker.service
```

```
Thu Sep  5 07:31:03 UTC 2024
sudo systemctl restart nomad.service
2024-09-05T07:32:06.624400+00:00 nomad-agent-terraform-1 systemd[1]: Stopping nomad.service - Nomad...
2024-09-05T07:32:11.738022+00:00 nomad-agent-terraform-1 systemd[1]: nomad.service: Main process exited, code=exited, status=1/FAILURE
2024-09-05T07:32:11.738146+00:00 nomad-agent-terraform-1 systemd[1]: nomad.service: Failed with result 'exit-code'.
2024-09-05T07:32:11.741856+00:00 nomad-agent-terraform-1 systemd[1]: Stopped nomad.service - Nomad.
2024-09-05T07:32:11.761288+00:00 nomad-agent-terraform-1 systemd[1]: Starting nomad.service - Nomad...
2024-09-05T07:32:13.958968+00:00 nomad-agent-terraform-1 systemd[1]: Started nomad.eligible.service - Nomad eligtion service. Wait until Nomad is up and make it eligible again.
2024-09-05T07:32:13.966919+00:00 nomad-agent-terraform-1 systemd[1]: Started nomad.service - Nomad.
2024-09-05T07:32:23.750174+00:00 nomad-agent-terraform-1 systemd[1]: nomad.eligible.service: Deactivated successfully.
```

```
Thu Sep  5 07:32:56 UTC 2024
sudo systemctl restart docker.service
2024-09-05T07:33:07.796698+00:00 nomad-agent-terraform-1 systemd[1]: Stopping docker.service - Docker Application Container Engine...
2024-09-05T07:33:07.798235+00:00 nomad-agent-terraform-1 systemd[1]: Stopping nomad.service - Nomad...
2024-09-05T07:33:07.815733+00:00 nomad-agent-terraform-1 systemd[1]: docker.service: Deactivated successfully.
2024-09-05T07:33:07.816080+00:00 nomad-agent-terraform-1 systemd[1]: Stopped docker.service - Docker Application Container Engine.
2024-09-05T07:33:07.822433+00:00 nomad-agent-terraform-1 systemd[1]: Starting docker.service - Docker Application Container Engine...
2024-09-05T07:33:08.651195+00:00 nomad-agent-terraform-1 systemd[1]: Started docker.service - Docker Application Container Engine.
2024-09-05T07:33:08.933513+00:00 nomad-agent-terraform-1 systemd[1]: nomad.service: Deactivated successfully.
2024-09-05T07:33:08.933703+00:00 nomad-agent-terraform-1 systemd[1]: Stopped nomad.service - Nomad.
2024-09-05T07:33:08.945481+00:00 nomad-agent-terraform-1 systemd[1]: Starting nomad.service - Nomad...
2024-09-05T07:33:10.232372+00:00 nomad-agent-terraform-1 systemd[1]: Started nomad.eligible.service - Nomad eligtion service. Wait until Nomad is up and make it eligible again.
2024-09-05T07:33:10.235648+00:00 nomad-agent-terraform-1 systemd[1]: Started nomad.service - Nomad.
2024-09-05T07:33:10.325235+00:00 nomad-agent-terraform-1 systemd[1]: nomad.eligible.service: Deactivated successfully.
```

</p>
</details>

Poseidon captures such occurrences and uses the runner as clean and idle.  

> The Nomad and Docker systemd services are connected via a systemd PartOf relationship.

Reviewer: Sure, but this is custom and not yet present in the docs, is it?

Author: I intended to introduce it with those two sentences 😅  Would you like to have a dedicated section in the docs?

Reviewer: I thought it might make sense to add a sample file `/etc/systemd/system/nomad.service.d/override.conf` to the repo (or inline in the docs) to have the explicit settings. Otherwise, one needs to figure out which service depends on which over and over again (see the uncertainty we just had in the other thread).

This results in Nomad being restarted once Docker restarts, but not vice versa.

> ### Nomad Event Stream

Reviewer: Should we move this section before the docs about Nomad Restarts? It would make sense, since you are referring to the Event Stream from within the docs about restarts.

The following table lists the behavior for restarts of Nomad Servers depending on its role and on whether Poseidon is connected to it (via DNS Resolution).

| Role     | DNS Resolves | | WebSocket Problem? | Event Stream Problem? | 
|----------|--------------|-|--------------------|-----------------------|
| Leader   | Yes          | | problematic        | problematic           |
| Leader   | No           | | problematic        | fine                  |
| Follower | Yes          | | problematic        | problematic           |
| Follower | No           | | fine               | fine                  |

Such restarts lead to problems with either individual WebSocket connections of executions or the Nomad Event Stream.
When the Nomad Event Stream is aborted, Poseidon tries to reestablish it. Once it succeeds in doing so, all environments and runners are recovered from Nomad.

In the case of Nomad Agent restarts, the WebSocket connection of a running execution aborts.
Furthermore, when also Docker of the Nomad Agent is restarted, the containers are recreated.
Poseidon captures such occurrences and uses the runner as clean and idle.

### Systemd Relationship Nomad - Docker

> The Nomad and Docker systemd services are connected via a systemd `PartOf` relationship.
> In a systemd overwrite (`/etc/systemd/system/nomad.service.d/override.conf`), the `PartOf` relationship is defined (`[Unit]
> PartOf = docker.service`).
> This results in Nomad being restarted once Docker restarts, but not vice versa.
> At the moment, it is necessary for Nomad to recreate the CNI network interfaces (See #490).

Reviewer: Since this line break is not shown in the rendered Markdown, let's simply use

````suggestion
We suggest to connect the Nomad and Docker systemd services through a systemd `PartOf` relationship.
In a systemd overwrite (`/etc/systemd/system/nomad.service.d/override.conf`), the `PartOf` relationship can be defined as follows:

```ini
# /etc/systemd/system/nomad.service.d/override.conf
[Unit]
PartOf = docker.service
```

This results in Nomad being restarted once Docker restarts, but not vice versa.
At the moment, this behavior is necessary so that Nomad can recreate the CNI network interfaces (see [#490](https://github.com/openHPI/poseidon/issues/490)).
````
